### ISSUE TITLE
Regressions pertaining to loading local pages and local resources in a UWP app

### ISSUE THREAD

**Issue by mqudsi** [MicrosoftEdge/WebView2Feedback#212]
The `WebView2` version that shipped with the February alpha preview of WinUI 3 could load local files via `file:///` syntax (in a UWP project) but it seems that the version included in the May 2020 WinUI Preview 1 can load neither `file:///` urls (in-browser exception stating access to local URLs is blocked, seemingly inherited from the electron codebase) nor the WebView 1.0 `ms-appx:///Assets/...` urls which seem to be completely unknown to WebView2 regardless of whether it's in a UWP or packaged WPF project. That's for setting the initial page.

Now as to media resources, while `NavigationStarted` can be used to intercept requests for actual pages, it doesn't fire for media resources. I *really* don't want to go back to using blobs delivered via shimmed `WebAllowedObjects` and a JS observer to mogrify links into JS events to load images and attachments, and it's going to be far worse because the original MS Edge had an override for the blob API that allowed automatic garbage collection of blob URIs after first use, but the Edgium equivalent API afaik doesn't, meaning all blob resource URLs will need to be manually collected after they've been loaded.

While exploring options when I found `.UriSource` missing (I later realized it was renamed to `.Source`), I discovered that this unexpectedly throws an `ExecutionEngineException`:

```csharp
webview.ExecuteScriptAsync($"window.location.replace(\"{url}\");");
```

Is there a story for loading HTML documents and their associated resources from a UWP app?

[AB#27073119](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/27073119)

**Comment by david-risney** [comment]
Thanks for trying out WebView2! There's a lot of great feedback here. Can you elaborate on your app scenario and how its using local content?

* **file URIs no longer working in UWP apps**: I'm not aware of this. I'll open a bug. Can you describe where the file URI is used? Is it in a Navigate call or as a image or video element source in another file URI?
* **ms-appx**: Yes, ms-appx / ms-appdata URIs are not yet supported and we have a feature request for that (#37).
* **NavigationStarting not firing for media resources**: NavigationStarting is only for top level document navigations, but we do have the WebResourceRequested event which covers all web resource resolutions that should cover media resources.
* **blob garbage collection**: We weren't planning on adding non-standard DOM APIs. Is using the standards based revokeObjectURL an issue?
* **ExecutionEngineException**: Not sure what's happening here. I'll open a bug and look into it.
* **Story for local resources in UWP app**: We have features to better support this in our backlog like custom URI schemes, ms-appx/ms-appdata URI schemes, and WinRT object injection into script. Without those features it should be harder to use local content in a UWP app, but we're working on it. Thanks!

**Comment by mqudsi** [comment]
Thanks for the response, @david-risney. My UWP app is "fully native" with the exception that it uses a webview as a widget to draw some locally generated/procured content (i.e. the webview does not communicate with the internet) using HTML and CSS.

> You said its for setting the initial page, so Is it in a Navigate call that you're using it and its blocked?

Yes, the first order of business is to load a local asset as the HTML page. The WebView2 control in WinUI Preview 1 does not have (or at least I don't see) a `Navigate()` method, only the `NavigateToString()` function and the `WebView2.Source` property, the latter of which fails to load a `file:///` url. Normally, I would then load some resources (css, js) via a URI relative to the `file:///` page itself; as I cannot get the `file:///` (or `ms-appx:///`) url to load, I cannot test whether or not it works for relative media resources.

> Can you describe where the file URI is used? You said its for setting the initial page, so Is it in a Navigate call that you're using it and its blocked?

Setting `WebView2.Url` to a `file:///` URL doesn't work and if you manually try to load any `file:///` content (say by opening the inspector and issuing `window.location.replace("file:///d:/path/to/index.html")` then Edgium throws the aforementioned exception inherited from the electron codebase "Not allowed to load a local resource"

![image](https://user-images.githubusercontent.com/606923/82935173-3eebd900-9f52-11ea-9dd2-5d7ffa89ab1e.png)

> NavigationStarting is only for top level document navigations, but we do have the WebResourceRequested event which covers all web resource resolutions that should cover media resources.

I'm not surprised that `NavigationStarting` doesn't fire for media URIs (it didn't with WebView 1.0 either); as for `WebResourceRequested`, I'd already come across references to it in the documentation/discussions, but there is no `WebResourceRequested` event exposed in this version of WinUI/WebView2 (using `WebView2Loader.dll` version `0.9.430.0`). Is this a C++-only API, available only in a later version of the library, or something else?

> We weren't planning on adding non-standard DOM APIs. Is using the standards based revokeObjectURL an issue?

As long as blob urls is not considered the primary way of resolving local/package media resources, then `revokeObjectUrl` is fine. I am glad to hear that non-standard DOM APIs are not being considered at this time *with that caveat*.

Re `ExecutionEngineException`:
> In your sample code above, what type is `url`?

A `System.Uri` object. I confirmed that this happens even with URLs that should be supported with this version, such as `new Uri("http://localhost:8000/index.html", UriKind.Absolute)`.

> We have features to better support this in our backlog like custom URI schemes, ms-appx/ms-appdata URI schemes, and WinRT object injection into script. Without those features it should be harder to use local content in a UWP app, but we're working on it. Thanks!

If the webview can resolve `file:///` URLs that are retrieved via sandbox-safe APIs (i.e. `Windows.Storage.StoreItem.Path` converted to a `file:///` uri), then `ms-appx:///` protocol support is just icing on the cake.

For WinRT object injection, [I wrote my own version of `AddWebAllowedObject` via `WebView2.WebMessageReceived` and `WebView2.PostMessage()`](https://gist.github.com/mqudsi/ceb4ecee76eb4c32238a438664783480) that has been working perfectly for our needs. It was actually a huge improvement in ergonomics for us because the lack of a pre-existing `AddWebAllowedObject()` meant that we weren't constrained by the extremely clunky semantics of the WebView 1 version of the same that required all parameters to be strings; instead we're just using `System.Text.JSON` and the JS `JSON` interface to marshall everything bidirectionally. The fact that the new `WebView2.ExecuteScriptAsync` doesn't just fire and run but actually waits for the execution to complete before posting the `IAsyncProgress` result means we don't need to use a separate `ManualResetEvent` or `SemaphoreSlim` to achieve lock-step execution between the C# backend and the JS frontend, so thank you *very* much for that bit of good interop news. `Task`s can now be marshalled to `Promise`s and vice-versa (although I can't find a clean way of detecting awaitable WinRT objects on the C# side and resort to reflection).

This is of course probably completely incompatible with .NET Native, but as thing currently stand, all of WinUI 3 isn't compatible anyway as it requires sideloading to work around the runtime `System.Private.CoreLib` dependency load that isn't declared in the UWP package and I have no idea if there are any plans to sort that out before final release or if non-side-loaded AppX packages are basically on their way out now.

**Comment by lingamy** [comment]
@mqudsi , thanks for your feedback! 

WebView2 can access location files using file protocol file:///. If you try it within a win32 app, it should work, please let us know otherwise.

For WinUI apps, as codendone has mentioned at https://github.com/microsoft/microsoft-ui-xaml/issues/1967, file:/// works, but only for files which the app has access to. UWP applications have restricted file system access, which is largely just the app package files and temp files. As a result, most files, including anything in C:\temp, would not be accessible. If you try to access a file within your UWP app package through file:///, do it work for you?

As david-risney mentioned above, ms-appx:/// support is in our backlog and should be available in a few months. We'll let you know once it's available.

**Comment by champnic** [comment]
We have improved support for local file access using the [SetVirtualHostNameToFolderMapping ](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2_3?view=webview2-1.0.781-prerelease&preserve-view=true#setvirtualhostnametofoldermapping)API available in 1.0.781-prerelease SDK. Please give it a try and let us know if it isn't working for you. Thanks!

**Comment by mqudsi** [comment]
I'm not able to test this because the version of WebView2 that is shipping with WinUI 3 does not expose the API in question.

However, from the entry point alone I know that isn't going to be sufficient: for paths that the user has manually selected via a FilePickerDialog (i.e. not part of the package itself) and granted the application access to, using basic Win32 APIs like `CreateFile` to access a file or folder in that path will fail due to the sandboxing. The path must be opened via the `Windows.Storage` API, referencing the previously saved location from the "favorites" list by correlating the path to the favorites list key in order for the call to succeed.

Via these design constraints foisted on everyone thanks to the limitations of Windows.Storage, the only API that would actually address the issues would resemble a callback returning a stream representing the result of the content load, or null to use the default location-based retrieval.
